File: demo_src/Fortran/2D/Gaussian/gaussian.f90
!* -------------------------------------------------------------------------- *
!* DESCRIPTION (reference phase where f(x,y) < 0):                            *
!* gaussian line in the square [0,1]x[0,1]                                    *
!* f(x,y) = y - yy0 - a0 exp[-ga (x - xx0)^2]                                 *
!* -------------------------------------------------------------------------- *

MODULE GAUSSIAN_MOD

  IMPLICIT NONE

  REAL(8), PARAMETER :: mypi = 3.141592653589793238462643D0
  REAL(8) :: yy0 = 0.220D0, a0 = 0.51D0
  REAL(8) :: xx0 = 0.541D0, ga = 60.3D0

  CONTAINS

!  initialize a random seed from the system clock at every run 
SUBROUTINE INIT_RANDOM_SEED()

      INTEGER :: i, n, clock
      INTEGER, DIMENSION(:), ALLOCATABLE :: seed
      INTRINSIC RANDOM_SEED, SYSTEM_CLOCK 

      CALL RANDOM_SEED(size = n)
      ALLOCATE(seed(n))

      CALL SYSTEM_CLOCK(COUNT=clock)

      seed = clock + 37 * (/ (i - 1, i = 1, n) /)
      CALL RANDOM_SEED(PUT = seed)

      DEALLOCATE(seed)
      
END SUBROUTINE INIT_RANDOM_SEED


SUBROUTINE INIT(randominput)  

  IMPLICIT NONE
  
  LOGICAL, INTENT(IN) :: randominput
  REAL(8) :: seed
  REAL(8), PARAMETER :: scalingfactor  = 0.06D0
    
  INTRINSIC RANDOM_NUMBER  

  IF(randominput .EQV. .TRUE.) THEN
  
    CALL INIT_RANDOM_SEED()

    ! yy0 --> from 0.19 to 0.25
    CALL RANDOM_NUMBER(seed) 
    yy0 = 0.19D0 + seed*scalingfactor
 
    ! xx0 --> from 0.511 to 0.561 
    CALL RANDOM_NUMBER(seed)
    xx0 = 0.511D0 + seed*scalingfactor

    ! a0 --> from 0.48 to 0.54
    CALL RANDOM_NUMBER(seed) 
    a0 = 0.48D0 + seed*scalingfactor;

    ! ga --> from 60.00 to 0.60.6 
    CALL RANDOM_NUMBER(seed)  
    ga = 60.D0 + seed*scalingfactor;

  END IF
  
END SUBROUTINE INIT


REAL(8) FUNCTION IMPL_FUNC(xyz)

  IMPLICIT NONE

  REAL(8), DIMENSION(*), INTENT(IN) :: xyz
  REAL(8) :: x,y

  INTRINSIC DEXP

  x = xyz(1)
  y = xyz(2)
    
  IMPL_FUNC = y - yy0 - a0*DEXP(-ga*(x-xx0)*(x-xx0))

  RETURN 

END FUNCTION IMPL_FUNC

!* -------------------------------------------------------------------------- *

SUBROUTINE CHECK_AREA(areanum, randominput)

  IMPLICIT NONE
  LOGICAL, INTENT(IN) :: randominput
  REAL(8), INTENT(IN) :: areanum
  REAL(8) :: areana

  ! analytical integration with x in [0,1] 
  areana = yy0 + 0.5*a0*SQRT(mypi/ga)*(ERF(SQRT(ga)*(1.-xx0) )-ERF(-SQRT(ga)*xx0));

  write(*,*) '-----------------------------------------------------------'
  write(*,*) '-------------------- F: gaussian check --------------------'
  write(*,100) areana
  write(*,101) areanum
  write(*,*) ' '
  write(*,102) DABS(areanum-areana)
  write(*,103) DABS(areanum-areana)/areana
  write(*,*) '-----------------------------------------------------------'
  IF(randominput .EQV. .FALSE.) THEN
    write(*,*) 'with Intel i7 3.4 GHz + Linux openSUSE 13.1 + gcc 4.8.1 -O2'
    write(*,*) '-----------------------------------------------------------'
    write(*,*) 'analytical area :  3.3640894546075423E-01'
    write(*,*) 'numerical  area :  3.3640894546075717E-01'
    write(*,*) ' '
    write(*,*) 'absolute error  :  2.9420910152566648E-15'
    write(*,*) 'relative error  :  8.7455790190926772E-15'
    write(*,*) '------------------ F: end gaussian check ------------------'
    write(*,*) '-----------------------------------------------------------'
  END IF
  write(*,*) ' '
  100 FORMAT(' analytical area : ', ES23.16)
  101 FORMAT(' numerical  area : ', ES23.16)
  102 FORMAT(' absolute error  : ', ES23.16)
  103 FORMAT(' relative error  : ', ES23.16)

END SUBROUTINE CHECK_AREA

END MODULE GAUSSIAN_MOD

File: demo_src/C/2D/Gaussian/gaussian.c
/****************************************************************************
 * Copyright (C) 2015 by Simone Bnà(a), Sandro Manservisi(a),               *
 * Ruben Scardovelli(a), Philip Yecko(b) and Stephane Zaleski(c,d)          *
 * (a) DIN–Lab. di Montecuccolino, Università di Bologna,                   *
 *     Via dei Colli 16, 40136 Bologna, Italy                               *
 * (b) Physics Department, Cooper Union, New York, NY, USA                  *
 * (c) Sorbonne Universités, UPMC Univ Paris 06, UMR 7190,                  *
 *     Institut Jean Le Rond d’Alembert, F-75005, Paris, France             *
 * (d) CNRS, UMR 7190, Institut Jean Le Rond d’Alembert, F-75005,           *
 *     Paris, France                                                        *
 *                                                                          *
 * You should have received a copy of the CPC license along with Vofi.      *
 * If not, see http://cpc.cs.qub.ac.uk/licence/licence.html.                *
 *                                                                          *
 * e-mail: ruben.scardovelli@unibo.it                                       *
 *                                                                          *
 ****************************************************************************/

#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <time.h>
#include "gaussian.h"

typedef const double creal;
typedef const int cint;
typedef double real;

/* -------------------------------------------------------------------------- *
 * DESCRIPTION (reference phase where f(x,y) < 0):                            *
 * gaussian line in the square [0,1]x[0,1]                                    *
 * f(x,y) = y - yy0 - a0 exp[-ga (x - xx0)^2]                                 *
 * -------------------------------------------------------------------------- */

static double yy0 = 0.22;
static double a0 = 0.51;
static double xx0 = 0.541;
static double ga = 60.3;


/* -------------------------------------------------------------------------- */

void init(cint randominput)
{

  if(randominput) {
    
    /* initialize random number generator. */
    srand(time(0));  
   
    double scalingfactor = 0.06;
    /* yy0 --> from 0.19 to 0.25 */
    yy0 = 0.19 + ((double)rand() / RAND_MAX)*scalingfactor;
  
    /* xx0 --> from 0.511 to 0.561 */
    xx0 = 0.511 + ((double)rand() / RAND_MAX)*scalingfactor;
  
    /* a0 --> from 0.48 to 0.54 */
    a0 = 0.48 + ((double)rand() / RAND_MAX)*scalingfactor;
  
    /* ga --> from 60.00 to 0.60.6 */
    ga = 60. + ((double)rand() / RAND_MAX)*scalingfactor;
  
  }
  
  return;
}

/* -------------------------------------------------------------------------- */

real impl_func(creal xy[]) 
{
  double x,y,f0;

  x = xy[0];
  y = xy[1];

  f0 = y - yy0 - a0*exp(-ga*(x-xx0)*(x-xx0));

  return f0;
}

/* -------------------------------------------------------------------------- */

void check_area(creal area_n, cint randominput)
{
  real area_a;

  /* analytical integration with x in [0,1]  */
  area_a = yy0 + 0.5*a0*sqrt(M_PI/ga)*(erf(sqrt(ga)*(1.-xx0) )-erf(-sqrt(ga)*xx0));
  
  fprintf (stdout,"----------------------------------------------------------------------\n");
  fprintf (stdout,"--------------------- C: gaussian check ------------------------------\n");
  fprintf (stdout," * gaussian line in the square [%.1f,%.1f]x[%.1f,%.1f] in a %dX%d grid   *\n", X0, X0+H, Y0, Y0+H, NMX, NMY);
  fprintf (stdout," * f(x,y) = y - yy0 - a0 exp[-ga (x - xx0)^2]                        *\n");
  fprintf (stdout,"----------------------------------------------------------------------\n");
  fprintf (stdout,"yy0:   %23.16e\n",yy0);
  fprintf (stdout,"xx0:   %23.16e\n",xx0);
  fprintf (stdout,"a0:    %23.16e\n",a0);
  fprintf (stdout,"ga:    %23.16e\n",ga);
  fprintf (stdout,"--------------------------------------------------------------------\n");
  fprintf (stdout,"analytical area : %23.16e\n",area_a);
  fprintf (stdout,"numerical  area : %23.16e\n\n",area_n);
  fprintf (stdout,"absolute error  : %23.16e\n",fabs(area_a-area_n));
  fprintf (stdout,"relative error  : %23.16e\n",fabs(area_a-area_n)/area_a); 
  fprintf (stdout,"-----------------------------------------------------------\n");
  if(!randominput) {
    fprintf (stdout,"with Intel i7 3.4 GHz + Linux openSUSE 13.1 + gcc 4.8.1 -O2\n");
    fprintf (stdout,"-----------------------------------------------------------\n");
    fprintf (stdout,"analytical area :  3.3640894546075428e-01\n");
    fprintf (stdout,"numerical  area :  3.3640894546075722e-01\n\n");
    fprintf (stdout,"absolute error  :  2.9420910152566648e-15\n");
    fprintf (stdout,"relative error  :  8.7455790190926756e-15\n");
    fprintf (stdout,"------------------ C: end gaussian check ------------------\n");
    fprintf (stdout,"-----------------------------------------------------------\n");
  }
  fprintf (stdout,"\n");
      
  return;
}
